src/cargo/ops/registry.rs

```diff
@@ -388,6 +388,14 @@ pub fn configure_http_handle(config: &Config, handle: &mut Easy) -> CargoResult<
     } else {
         handle.useragent(&version().to_string())?;
     }
+
+    let noproxy = env::var("NO_PROXY").ok()
```

Comment:
I'm not sure I understand the original motivation for this.  It looks like libcurl reads these environment variables automatically (https://github.com/curl/curl/blob/835a2fe694214341b601b2439082a30a24cc9da9/lib/url.c#L2707-L2723) and has for a while (though it looks like its been changed a bit over the years).  Testing on my system, stable cargo reads them just fine.  I have libcurl 7.54 which isn't exactly new.

What exactly is the issue here?  I'm concerned without understanding why it doesn't work for the original reporter that this might not fix it for them.  Maybe they have an old version of libcurl?  Or a misconfiguration?

Reply:
I am the original reporter, and I have tried to get the variable to work for me, but it just doesn't. Neither on my mac, or in a CI environment with CentOS 7. Here is the version of libraries that Cargo is using:

```
λ otool -L ~/.cargo/bin/cargo
/Users/nate/.cargo/bin/cargo:
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 57740.60.18)
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.8)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1349.8.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
```

Reply:
You know what, I may have been using lowercased `no_proxy`. Let me try tomorrow with `NO_PROXY`

Reply:
@natemara any luck with tests to see how it went?

Reply:
@natemara what version of libcurl do you have? (`curl -V` should be enough I think)

Reply:
I posted above the version of libcurl that is linked to cargo, but 

```
λ /usr/bin/curl -Vcurl 7.54.0 (x86_64-apple-darwin16.0) libcurl/7.54.0 SecureTransport zlib/1.2.8
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz UnixSockets
```

Reply:
I don't see how the older code would have a problem.  It's the exact same version I have on my system, and it seems to work fine for me.  The curl change ([1919](https://github.com/curl/curl/pull/1919)) looks like it was just a code cleanup.  There was a significant fix ([1140](https://github.com/curl/curl/pull/1140)) that is relevant to this, but that was in 7.53 (that's why I wanted to know the exact version).

I'm not opposed to this change, I'm just not convinced it will fix the issue.

I'm wondering if there's a simple way to test curl on the command-line against @natemara's git server?  Or if @natemara is willing to try to test this PR (not sure how difficult that would be).  If you are, the steps would roughly be:
1. Checkout cargo from this repo.
2. Pull this PR (`git fetch origin pull/6115/head:noproxy`)
3. `cargo build` (generally should work in most environments)
4. Go to your project with the problem and run `/path/to/cargo/target/debug/cargo fetch` with the appropriate proxy environment variables.

Alternatively we can just merge this and wait for it to hit nightly.

Reply:
I followed those instructions, and it appears that this does not solve my problem.

Reply:
Ah yes you're right @ehuss, I was misreading that, my bad! @natemara are you sure that the version of Cargo you compiled includes these changes? I think @ehuss's instructions are missing a git checkout step (it looks like they're just fetching), so if you followed them exactly then youmay have just built master cargo

(I think you can fix it by injecting a `git checkout noproxy`)

Reply:
I did do the `git checkout noproxy`, and I have been looking at the cargo source to get more info, and I can definitely confirm that my source has your noproxy changes.

Reply:
Ah, yes, my apologies.

Reply:
Ok thanks for confirming, I don't think I understand though what you mean when you say Cargo goes through different flows. 

Cargo does have a path where it detects a proxy is in use so it configures libgit2 to use those proxy settings (by using our libcurl handles), but I'm not sure how that's related here. When cargo creates a curl handle, with this PR, it should be appropriately configured to avoid using proxies for particular hosts. Even moreso from what @ehuss is mentioning it shouldn't even need this PR actually, so long as curl is used to perform a git fetch no change to Cargo should be necessary to pick up `NO_PROXY` configuration.

@natemara are you sure you're configuring `NO_PROXY` in a way that curl understands? For example I get [this transcript today](https://gist.github.com/alexcrichton/f9a576fc4a493e9db41f0a61434cb4c6) with cargo build from master, which seems to indicate that `NO_PROXY` is indeed being read and respected.

Reply:
To me it just doesn't look like the upload-pack stuff is happening in the trace with no proxy settings, and that's the step that fails with the proxy settings.

Reply:
I did some more testing while watching the HTTP logs on the git server. When using an HTTP proxy, even when my git server is the only host in the `no_proxy` variable, cargo behaves differently.

With `http_proxy`, `no_proxy`, etc. set:

```
﻿October 8th 2018, 11:26:06.000/tools/logger/info/refs?service=git-upload-pack GET 301
October 8th 2018, 11:26:07.000/tools/logger.git/info/refs?service=git-upload-pack GET 200
October 8th 2018, 11:26:08.000/tools/logger.git/info/refs?service=git-upload-pack GET 200
October 8th 2018, 11:26:08.000/tools/logger/git-upload-pack POST 404
```

With no env vars set:

```
October 8th 2018, 11:32:30.000/tools/logger/info/refs?service=git-upload-pack GET 301
October 8th 2018, 11:32:31.000/tools/logger.git/git-upload-pack POST 200
October 8th 2018, 11:32:31.000/tools/logger.git/info/refs?service=git-upload-pack GET 200
```

With the env vars, cargo sends a POST to `.../logger/git-upload-pack`. With no vars, the same request is a POST to `.../logger.git/git-upload-pack`. That's why cargo get's a 404 back, it's sending the request to a different URL.

Reply:
I believe GitHub transparently handles `/foo` vs `/foo.git`, is this a case where the url listed in Cargo.toml needs a trailing `.git` for your server?

Reply:
Adding `.git` should fix it, but that is chopped off at https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/git/source.rs#L102.

Reply:
And yes, I believe our git server (gitlab) requires the `.git` at the end.

Reply:
If I comment out the block that I linked above, it works perfectly.

Reply:
@natemara do you have other dependencies on the same crate elsewhere? That canonicalization is purely used for hashing and comparison, it's not actually used for fetching dependencies. Instead what's written down exactly in `Cargo.toml` is used to fetch dependencies. If you have one dependency on `/logger` and one on `/logger.git` (for example) then they'll has to the same bucket and resolve to the same dependency (and sort of nondeterministically pick one of the dependencies to fetch).

I don't really understand how commenting out the logic would fix anything because we don't use the "canonical URL" to actually fetch anything. Do you know of any other dependencies active throughout the system on this `/logger.git` url?

Reply:
My mistake - when I got it to work I had the proxy settings disabled. I think I've made a small example of my problem that can be distributed so that you can see the behavior for yourself. If you download this project https://gitlab.com/natemara/cargo-proxy-consumer, then run the `run` script in the root dir, you should be able to download the dependency just fine. If you uncomment lines 10-15 and run it again, the download should fail with 404s.

Were you able to reproduce this?